Adafruit_DAP_SAM.cpp

```diff
@@ -128,13 +130,101 @@ bool Adafruit_DAP_SAM::select(uint32_t *found_id) {
     if (device->dsu_did == DAP_DSU_did) {
       target_device = *device;
 
+      locked = dap_read_word(DAP_DSU_CTRL_STATUS) & 0x00010000;
+      if (locked) {
+        Serial.println("Device is locked, must be unlocked!");
+      }
+      else {
+        // Stop the core
+        finishReset();
+      }
       return true;
     }
   }
 
   return false;
 }
 
+void Adafruit_DAP_SAM::finishReset() {
+	// Stop the core
+	dap_write_word(DHCSR, 0xa05f0003);
+	dap_write_word(DEMCR, 0x00000001);
+	dap_write_word(AIRCR, 0x05fa0004);
+}
+
+void Adafruit_DAP_SAM::programFlash(const uint8_t * source, uint16_t bootloaderSize, bool doVerify, uint32_t flashOffset) {
+  size_t bufSize = pageSize();
+  uint8_t verifyBuf[bufSize];
+
+  Serial.println("\nProgramming...");
+  uint32_t startAddr = program_start(flashOffset);
+  uint32_t addr = startAddr;
+  while (addr < bootloaderSize) {
+    Serial.print(" 0x");
+    Serial.print(addr, HEX);
+    programBlock(addr, source + addr);
+    addr += bufSize;
+  }
+
+  if (doVerify) {
+    bool success = true;
+    Serial.println("\nVerifying...");
+
+    addr = startAddr;
+    while (addr < bootloaderSize) {
+      Serial.print("Block: 0x");
+      Serial.print(addr, HEX);
+      Serial.print(" ");
+      readBlock(addr, verifyBuf);
+      if (memcmp(verifyBuf, source + addr, bufSize) == 0) {
+        Serial.println("OK");
+      }
+      else {
+        Serial.println("Failed compare!");
+        success = false;
+      }
+      addr += bufSize;
+    }
+
+    if (!success) {
+      perror_exit("Error validating ");
+    }
+  }
+}
+
+void Adafruit_DAP_SAM::resetProtectionFuses(bool resetBootloaderProtection, bool resetRegionLocks) {
```

Comment:
should be moved to user sketch, the DAP API should be keep as minimal to primitive dap protocol only.

Reply:
There is already a mix of high-level and low level methods sprinkled around, and I think it is good if there is methods that show how thigs should be done and could be of use to someone. It is important to limit flash erase/write operations because of wear, so it important for a library such as this to in some way, show methods for how to do it. My code may not be the best example, but something along the same lines. Resetting these parameters could easily be required for being able to successfully flash a micro controller like the SAMD21. I see it isn't strictly necesary for the SAMD51, but certainly is for the SAMD21.

The amount of repetetive code in the sketches is amazingly large, and it seemed appropriate for the Adafruit_DAP_SAM::program_start method to reset the protection fuses when necessary, since the use of Adafruit DAP is to actually flash new bootloader (or any code really) into the microcontroller, and if the protection fuses are set, the flashing will fail. 

The trouble with your approach is that every sketch using the API needs to be exapnded with more code, to do it correctly.

Reply:
I have to modify my standing a bit: Adafruit_DAP_SAM::program_start(uint32_t offset) now calls resetProtectionFuses(true, false) every time. 
This is only necessary if the offset is in the flash memory area protected by the BOOTPROT fuse.